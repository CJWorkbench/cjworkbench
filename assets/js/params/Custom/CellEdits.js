import React from 'react'
import PropTypes from 'prop-types'
import { withJsonStringValues } from '../util'

export class CellEdits extends React.Component {
  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.shape({
      col: PropTypes.string.isRequired,
      edits: PropTypes.arrayOf(PropTypes.shape({
        row: PropTypes.number.isRequired,
        value: PropTypes.string.isRequired
      }).isRequired)
    }).isRequired).isRequired,
  }

  // Starting from a JSON array like this
  // [
  //   { row: 3, col: 'foo', value:'bar' },
  //   { row: 6, col: 'food', value:'sandwich' },
  //   ...
  // ]
  // convert it to one that combines all edits for single column, like this
  // [
  //   { col: 'foo', edits: [ { row: 3, value: 'bar'}, ... ] },
  //   { col: 'food', edits: [ { row: 6, value: 'sandwich}, ... ] },
  // ]
  get editList () {
    const { value } = this.props
    const edits = {}

    for (const edit of value) {
      const colName = edit.col
      const colEdits = edits[colName] || []
      colEdits.push({ row: edit['row'], value: edit['value']})
      edits[colName] = colEdits
    }

    // Sort column names and convert to array
    const colNames = Object.keys(edits)
    colNames.sort()

    const editsArray = colNames.map(colName => {
      return { col: colName, edits: edits[colName] }
    })

    return editsArray
  }


  render() {
    return  (
      <React.Fragment>
        {this.editList.map(({ col, edits }) => (
          <div className='cell-edits--column' key={col}>
            <div className='edited-column'>{col}</div>
            <div className='cell-edits--rows'>
              {edits.map(({ row, value }, idx) => (
                <div className='cell-edits--row' key={idx}>
                  <div className='edited-row'>{row + 1}</div>
                  <div className='edited-text'>{value}</div>
                </div>
              ))}
            </div>
          </div>
        ))}
      </React.Fragment>
    )
  }
}

export default withJsonStringValues(CellEdits, [])
