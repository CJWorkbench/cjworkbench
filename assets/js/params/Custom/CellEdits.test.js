/* globals describe, expect, it */
import CellEdits from './CellEdits'
import { mount } from 'enzyme'

describe('CellEdits', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <CellEdits
        value={[
          { row: 3, col: 'foo', value: 'bar' },
          { row: 6, col: 'food', value: 'sandwich' },
          { row: 5, col: 'foo', value: 'gak' },
          { row: 17, col: 'food', value: 'pizza' }
        ]}
      />
    )

    expect(wrapper.find('.cell-edits--column')).toHaveLength(2)
    expect(wrapper.find('.cell-edits--row')).toHaveLength(4)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders empty data', () => {
    const wrapper = mount(
      <CellEdits value={[]} />
    )

    expect(wrapper.find('.cell-edits--column')).toHaveLength(0)
    expect(wrapper.find('.cell-edits--row')).toHaveLength(0)

    expect(wrapper).toMatchSnapshot()
  })
})
