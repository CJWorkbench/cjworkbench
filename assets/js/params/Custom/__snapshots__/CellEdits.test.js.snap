// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellEdits renders correctly 1`] = `
<CellEdits
  value={
    Array [
      Object {
        "col": "foo",
        "row": 3,
        "value": "bar",
      },
      Object {
        "col": "food",
        "row": 6,
        "value": "sandwich",
      },
      Object {
        "col": "foo",
        "row": 5,
        "value": "gak",
      },
      Object {
        "col": "food",
        "row": 17,
        "value": "pizza",
      },
    ]
  }
>
  <div
    className="cell-edits--column"
    key="foo"
  >
    <div
      className="edited-column"
    >
      foo
    </div>
    <div
      className="cell-edits--rows"
    >
      <div
        className="cell-edits--row"
        key="0"
      >
        <div
          className="edited-row"
        >
          4
        </div>
        <div
          className="edited-text"
        >
          bar
        </div>
      </div>
      <div
        className="cell-edits--row"
        key="1"
      >
        <div
          className="edited-row"
        >
          6
        </div>
        <div
          className="edited-text"
        >
          gak
        </div>
      </div>
    </div>
  </div>
  <div
    className="cell-edits--column"
    key="food"
  >
    <div
      className="edited-column"
    >
      food
    </div>
    <div
      className="cell-edits--rows"
    >
      <div
        className="cell-edits--row"
        key="0"
      >
        <div
          className="edited-row"
        >
          7
        </div>
        <div
          className="edited-text"
        >
          sandwich
        </div>
      </div>
      <div
        className="cell-edits--row"
        key="1"
      >
        <div
          className="edited-row"
        >
          18
        </div>
        <div
          className="edited-text"
        >
          pizza
        </div>
      </div>
    </div>
  </div>
</CellEdits>
`;

exports[`CellEdits renders empty data 1`] = `
<CellEdits
  value={Array []}
/>
`;
